Firestore/core/src/firebase/firestore/remote/stream_operation.cc

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/stream_operation.h"
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using util::AsyncQueue;
+
+StreamOperation::StreamOperation(GrpcStream* stream,
+                                 grpc::GenericClientAsyncReaderWriter* call,
+                                 AsyncQueue* firestore_queue)
+    : observer_{stream}, call_{call}, firestore_queue_{firestore_queue} {
+}
+
+void StreamOperation::UnsetObserver() {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  observer_ = nullptr;
+}
+
+void StreamOperation::Execute() {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  DoExecute(call_);
+}
+
+void StreamOperation::WaitUntilOffQueue() {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  if (!off_queue_future_.valid()) {
+    off_queue_future_ = off_queue_.get_future();
+  }
+  return off_queue_future_.wait();
+}
+
+std::future_status StreamOperation::WaitUntilOffQueue(
+    std::chrono::milliseconds timeout) {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  if (!off_queue_future_.valid()) {
+    off_queue_future_ = off_queue_.get_future();
+  }
+  return off_queue_future_.wait_for(timeout);
+}
+
+void StreamOperation::Complete(bool ok) {
+  // This mechanism allows `GrpcStream` to know when the operation is off the
+  // GRPC completion queue (and thus this operation no longer requires the
+  // underlying GRPC objects to be valid).
+  off_queue_.set_value();
```

Comment:
It seems like the way we're using these promises we could replace them with a semaphore.

Reply:
You have a `conditional_variable` in mind, right? I think when we discussed this originally, the idea was that a conditional variable would be lower level than necessary here. I don't feel strongly about it, though.